crates/vm2/src/instruction_handlers/heap_access.rs

```diff
@@ -153,11 +148,9 @@ pub(crate) fn grow_heap<T, W, H: HeapFromState>(
     state: &mut State<T, W>,
     new_bound: u32,
 ) -> Result<(), ()> {
-    let already_paid = H::get_heap_size(state);
-    if *already_paid < new_bound {
-        let to_pay = new_bound - *already_paid;
-        *already_paid = new_bound;
+    if let Some(to_pay) = new_bound.checked_sub(*H::get_heap_size(state)) {
         state.use_gas(to_pay)?;
+        *H::get_heap_size(state) = new_bound;
```

Comment:
Should this go before or after the `use_gas()` call? (The ordering matters since the call short-circuits on an error.) AFAIU, the previous logic changed the heap bound before the call – was it a part of the buggy behavior?

Reply:
It is important that it comes after so size only grows if it can be paid for.